packages/mobile/src/services/offline-downloader/offline-downloader.ts

```diff
@@ -0,0 +1,102 @@
+import path from 'path'
+
+import type { Track, UserMetadata } from '@audius/common'
+import {
+  encodeHashId,
+  accountSelectors,
+  cacheTracksSelectors
+} from '@audius/common'
+import { uniq } from 'lodash'
+import RNFS, { exists } from 'react-native-fs'
+
+import { store } from 'app/store'
+
+import { apiClient } from '../audius-api-client'
+
+import {
+  getLocalAudioPath,
+  getLocalCoverArtPath,
+  getLocalTrackJsonPath
+} from './offline-storage'
+const { getUserId } = accountSelectors
+const { getTrack } = cacheTracksSelectors
+
+// Main entrypoint - perform all steps required to complete a download
+export const downloadTrack = async (trackId: number, collection: string) => {
+  const state = store.getState()
+  const track = getTrack(state, { id: trackId })
+  if (!track) return false
+
+  const coverArtUri = getCoverArtUri(track)
+  const [coverArtDirectory, coverArtFileName] = getLocalCoverArtPath(track)
+  if (coverArtUri && coverArtDirectory && coverArtFileName) {
+    await downloadIfNotExists(coverArtUri, coverArtDirectory, coverArtFileName)
+  }
+  await tryDownloadTrackFromEachCreatorNode(track)
+  await writeTrackJson(track, collection)
+}
+
+// Unlike mp3 and album art, here we overwrite even if the file exists to ensure we have the latest
+const writeTrackJson = async (track: Track, collection: string) => {
+  const trackToWrite: Track = {
+    ...track,
+    offline: {
+      downloaded_from_collection: uniq([
+        collection,
+        ...(track?.offline?.downloaded_from_collection ?? [])
+      ]),
+      download_completed_time: Date.now(),
+      last_verified_time: Date.now()
+    }
+  }
+
+  const pathToWrite = getLocalTrackJsonPath(track)
+  await RNFS.write(pathToWrite, JSON.stringify(trackToWrite))
+}
+
+const tryDownloadTrackFromEachCreatorNode = async (track: Track) => {
+  const state = store.getState()
+  const user = (
```

Comment:
Do we not already have this in the store?

Reply:
Oh, we probably do. I was teaching myself how to use the API 😅

Reply:
Actually, wait. Not necessarily. If we're downloading all of our favorites we may not have loaded all of them yet. I think I have a mix of logic meant to work in the background and logic that doesn't. It's going to need some further separation and cleanup.